### ISSUE TITLE
Use `fmt::format` in logging

### ISSUE THREAD

**Issue by andrerenard** [kotekan/kotekan#466]
The new `fmt` formatting library seems to be the way of the future (to be included in C++20).  We should move the `ERROR|WARN|INFO|DEBUG|...` logging calls to use it instead of the current `printf` standard.   This is likely a huge project, since we need to update all the existing logging calls.  However if we do this, then it'll mark the end of needing `.c_str()`, and funny macros for 64-bit ints all the time.

**Comment by andrerenard** [comment]
One thing to watch out for here, there are a few places, `buffers.c` being the main one, that still use C, so won't work with `fmt` includes.  For those likely the best thing to do is to rename the existing ones something like `ERRORF|WARNF`, etc.

**Comment by nritsche** [comment]
There are two implementations of `internal_logging()`, a C-one in `errors.c` and a C++-one in `kotekanLogging.cpp`. As far as I understand, I can just replace the implementation in `kotekanLogging.cpp`, so kotekan's C code will use the unchanged logging functions in `errors.c`. There would be no need to rename the macros for use in non-C++-code, they just won't support anything but printf-style formatting there.

**Comment by andrerenard** [comment]
Ah you are right, that should work.

**Comment by nritsche** [comment]
The `fmt` library also supports printf-style formatting. But we really don't only want to swap out the underlying library but also the format style, right?

**Comment by nritsche** [comment]
> Ah you are right, that should work.

I will only have to make some classes inherit from `kotekanLogging`, otherwise they use the C-implementation of `internal_logging()`.

**Comment by nritsche** [comment]
Now that I am making everything use `kotekanLogging` as a base class, I find that any `static` methods can't use the logging macros, because they would check against the member variable that keeps the log_level.

**Comment by nritsche** [comment]
I finally decided that this code is completely unreadable if it uses the same names on different subscopes and for logging using printf and python style. The macros do different things depending on if they are used in an object that inherited from `kotekan::kotekanLogging`, if they are used in a static method or in C code.
Therefor I am renaming stuff now, creating macros with suffixes `_NON_OO` and `_F` and also I am splitting up `__log_level` into `_global_log_level` and `_member_log_level`.
Isn't the double underscore reserved for compiler internals? @andrerenard

**Comment by nritsche** [comment]
`fmt` offers compile-time syntax checking. But I couldn't figure out a way to make this work for strings that are getting reused, e.g.:
```
std::string my_string = fmt::format("This is bad: {:s}", e.what());
ERROR(my_string);
reply_error(my_string);
```

I think having compile-time format syntax would be great (I already found a few buggy log lines). But we wouldn't be able to reuse strings like above, which adds some redundant lines here and there.

**Comment by andrerenard** [comment]
Sorry I keep forgetting to comment on this.  I agree with the change to suffixes `_NON_OO` and `_F` and changing the `log_level` variable names.   

I'm not too sure I follow the compile time syntax checking issue.  What is it about `fmt` that prevents reusing the string like that?  Does it have to be within the macro to work?

**Comment by nritsche** [comment]
> What is it about `fmt` that prevents reusing the string like that? Does it have to be within the macro to work?
The macro for compile-time checks (`fmt()`) doesn't accept anything but string literals.
If we want both, the possibility to pass a string (not a string literal) to the logging macros and also compile-time checks with the `fmt()` macro, we have to write some clever macro logic that only does the compile-time check depending on the type.

**Comment by nritsche** [comment]
Another argument against compile-time checks is that `fmt::format` would has to be called in the macro. Which means we can't use `fmt::format_args` (see the bottom of https://github.com/fmtlib/fmt#examples) and I understand that this will bloat code size and compile time.

I compiled with the checks in place now and fixed all formatting mistakes we had in the code base / after I changed the format style.
